File: tests/kotlin/test_7_4_1_merge_steps.kt
package org.apache.tinkerpop.gremlin.tinkercat.tests

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import org.apache.tinkerpop.gremlin.structure.Vertex
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerCat
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerVertex

/**
 * Tests for Task 7.4.1: mergeV() and mergeE() Upsert Steps.
 *
 * TinkerCat's Kotlin API does not expose built-in mergeV/mergeE methods.
 * These tests implement the upsert semantics inline with Kotlin sequences
 * and verify the behaviour is correct — establishing what a future built-in
 * implementation must satisfy.
 * Mirrors tests/python/test_7_4_1_merge_steps.py.
 */
class Test_7_4_1_MergeSteps : StringSpec({

    lateinit var graph: TinkerCat

    beforeTest { graph = TinkerCat.open() }

    // ---------------------------------------------------------------------------
    // Inline mergeV / mergeE helpers (no TinkerCat built-in yet)
    // ---------------------------------------------------------------------------

    fun mergeV(
        label: String,
        criteria: Map<String, Any?>,
        onCreate: Map<String, Any?> = criteria,
        onMatch: Map<String, Any?> = emptyMap(),
    ): Vertex {
        val candidate = graph.vertices().asSequence()
            .filter { it.label() == label }
            .filter { criteria.all { (k, v) -> it.value<Any?>(k) == v } }
            .firstOrNull()
        return if (candidate != null) {
            onMatch.forEach { (k, v) -> (candidate as TinkerVertex).property(k, v) }
            candidate
        } else {
            val args = mutableListOf<Any?>(TinkerVertex.T.label, label)
            onCreate.forEach { (k, v) -> args += k; args += v }
            graph.addVertex(*args.toTypedArray())
        }
    }

    fun mergeE(
        label: String,
        outV: Vertex,
        inV: Vertex,
        criteria: Map<String, Any?> = emptyMap(),
        onMatch: Map<String, Any?> = emptyMap(),
    ) = graph.edges().asSequence()
        .filter { it.label() == label && it.outVertex().id() == outV.id() && it.inVertex().id() == inV.id() }
        .filter { criteria.all { (k, v) -> it.value<Any?>(k) == v } }
        .firstOrNull()
        ?.also { e -> onMatch.forEach { (k, v) -> e.property(k, v) } }
        ?: run {
            val args = mutableListOf<Any?>()
            criteria.forEach { (k, v) -> args += k; args += v }
            outV.addEdge(label, inV, *args.toTypedArray())
        }

    // ---------------------------------------------------------------------------
    // mergeV tests
    // ---------------------------------------------------------------------------

    "merge_v creates vertex when none matches" {
        val v = mergeV("person", mapOf("email" to "alice@example.com"),
            onCreate = mapOf("email" to "alice@example.com", "created" to true))
        graph.vertices.size shouldBe 1
        v.value<String>("email") shouldBe "alice@example.com"
        v.value<Boolean>("created") shouldBe true
    }

    "merge_v matches existing vertex without duplication" {
        val existing = graph.addVertex(TinkerVertex.T.label, "person",
            "email", "alice@example.com", "lastSeen", "old") as TinkerVertex
        mergeV("person", mapOf("email" to "alice@example.com"),
            onMatch = mapOf("lastSeen" to "new"))
        graph.vertices.size shouldBe 1
        existing.value<String>("lastSeen") shouldBe "new"
    }

    "merge_v with no on_create uses search criteria" {
        val v = mergeV("person", mapOf("name" to "Bob"))
        v.value<String>("name") shouldBe "Bob"
    }

    "merge_v with no on_match returns vertex unchanged" {
        graph.addVertex(TinkerVertex.T.label, "person", "email", "x@y.com", "score", 10)
        val v = mergeV("person", mapOf("email" to "x@y.com"))
        v.value<Int>("score") shouldBe 10
    }

    "merge_v is idempotent" {
        val criteria = mapOf("email" to "alice@example.com")
        mergeV("person", criteria)
        mergeV("person", criteria)
        graph.vertices.size shouldBe 1
    }

    // ---------------------------------------------------------------------------
    // mergeE tests
    // ---------------------------------------------------------------------------

    "merge_e creates edge when none exists" {
        val a = graph.addVertex(TinkerVertex.T.label, "person")
        val b = graph.addVertex(TinkerVertex.T.label, "person")
        mergeE("knows", a, b)
        graph.edges.size shouldBe 1
    }

    "merge_e matches existing edge without duplication" {
        val a = graph.addVertex(TinkerVertex.T.label, "person")
        val b = graph.addVertex(TinkerVertex.T.label, "person")
        a.addEdge("knows", b, "since", "2020")
        mergeE("knows", a, b, onMatch = mapOf("since" to "2024"))
        graph.edges.size shouldBe 1
        graph.edges().asSequence().first().value<String>("since") shouldBe "2024"
    }

    "merge_e is idempotent" {
        val a = graph.addVertex(TinkerVertex.T.label, "n")
        val b = graph.addVertex(TinkerVertex.T.label, "n")
        mergeE("link", a, b)
        mergeE("link", a, b)
        graph.edges.size shouldBe 1
    }

    "merge_e result has correct label" {
        val a = graph.addVertex(TinkerVertex.T.label, "n")
        val b = graph.addVertex(TinkerVertex.T.label, "n")
        val e = mergeE("likes", a, b)
        e.label() shouldBe "likes"
    }
})

File: tests/python/test_7_4_1_merge_steps.py
"""
Tests for Task 7.4.1: mergeV() and mergeE() Upsert Steps.

All tests call g.traversal().merge_v() / merge_e() directly.
These spawning steps are not yet implemented; every test fails with
AttributeError until they are added to TinkerCat's traversal API.
See docs/project/changelog/task-7.4.1-merge-steps.adoc.
"""

import pytest

from tinkercat import TinkerCat


@pytest.fixture
def g():
    graph = TinkerCat()
    yield graph
    graph.close()


# ---------------------------------------------------------------------------
# mergeV tests
# ---------------------------------------------------------------------------

def test_merge_v_creates_when_missing(g):
    criteria = {"label": "person", "email": "alice@example.com"}
    on_create = {"label": "person", "email": "alice@example.com", "created": True}
    result = g.traversal().merge_v(criteria, on_create=on_create).to_list()  # AttributeError
    assert len(result) == 1
    assert g.vertex_count == 1
    assert result[0].value("email") == "alice@example.com"
    assert result[0].value("created") is True


def test_merge_v_matches_existing(g):
    g.add_vertex("person", email="alice@example.com", lastSeen="old")
    result = g.traversal().merge_v(  # AttributeError
        {"label": "person", "email": "alice@example.com"},
        on_match={"lastSeen": "new"},
    ).to_list()
    assert g.vertex_count == 1
    assert result[0].value("lastSeen") == "new"


def test_merge_v_no_on_create_uses_search_criteria(g):
    result = g.traversal().merge_v({"label": "person", "name": "Bob"}).to_list()  # AttributeError
    assert result[0].value("name") == "Bob"


def test_merge_v_no_on_match_returns_unchanged(g):
    g.add_vertex("person", email="x@y.com", score=10)
    result = g.traversal().merge_v({"label": "person", "email": "x@y.com"}).to_list()  # AttributeError
    assert result[0].value("score") == 10


def test_merge_v_idempotent(g):
    criteria = {"label": "person", "email": "alice@example.com"}
    g.traversal().merge_v(criteria).iterate()  # AttributeError
    g.traversal().merge_v(criteria).iterate()
    assert g.vertex_count == 1


# ---------------------------------------------------------------------------
# mergeE tests
# ---------------------------------------------------------------------------

def test_merge_e_creates_edge(g):
    a = g.add_vertex("person")
    b = g.add_vertex("person")
    result = g.traversal().merge_e({"label": "knows", "out": a, "in": b}).to_list()  # AttributeError
    assert g.edge_count == 1
    assert result[0].label == "knows"


def test_merge_e_matches_existing_edge(g):
    a = g.add_vertex("person")
    b = g.add_vertex("person")
    g.add_edge("knows", a, b, since="2020")
    result = g.traversal().merge_e(  # AttributeError
        {"label": "knows", "out": a, "in": b},
        on_match={"since": "2024"},
    ).to_list()
    assert g.edge_count == 1
    assert result[0].value("since") == "2024"


def test_merge_e_requires_label(g):
    a = g.add_vertex("n")
    b = g.add_vertex("n")
    with pytest.raises(Exception):
        g.traversal().merge_e({"out": a, "in": b}).iterate()  # AttributeError or ValueError


def test_merge_e_idempotent(g):
    a = g.add_vertex("n")
    b = g.add_vertex("n")
    g.traversal().merge_e({"label": "link", "out": a, "in": b}).iterate()  # AttributeError
    g.traversal().merge_e({"label": "link", "out": a, "in": b}).iterate()
    assert g.edge_count == 1
